### ISSUE TITLE
More fine-grained tracking of infective options

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#5417]
The impact of 7bc919262a0b35e9ea78d988b282b3a20edf1c75 (re #4908) on the standard-library may be bigger than expected, see https://github.com/agda/agda-stdlib/issues/1509, so we should postpone this change until after the release, imo.

(The general rule: do not rush features into the release last minute...)

**Comment by jespercockx** [comment]
@andreasabel thanks for the feedback. In general I agree that we should not rush new features just before the release. However, in this case I believe I had a good reason to make an exception, so let me try to explain my decision:

First, we had a long discussion about this in https://github.com/agda/agda/issues/4908, where @nad brought up a (in my opinion) good argument for disabling these two flags by default. I tried out a few of the alternatives, including required changes on the std-lib which was quite a bit of (boring) work. I then asked for feedback from you, @UlfNorell , and @nad . I only got feedback from @nad who said the impact on the std-lib was not too bad, and the maintainers of the std-lib didn't raise any serious complaints on my PR either. Since the release was getting closer and this was one of the few remaining open issues, I went ahead and merged in the changes so the release could continue.

If you think I shouldn't have merged the PR, please let me know where exactly I went wrong. I would also still be grateful to hear your opinion on the other alternative solutions that I tried out.

**Comment by nad** [comment]
@andreasabel, is the problem that you are now asked to turn on `--guardedness` in code that imports code from the standard library, and that did not use to be marked as `--safe`? This can be fixed by adding `--guardedness` to the `flags` field of your `.agda-lib` files. I don't really see the problem with declaring explicitly that your code depends on code that uses `--guardedness`.

**Comment by andreasabel** [comment]
@nad:
> I don't really see the problem with declaring explicitly that your code depends on code that uses `--guardedness`.

Well, my code does not depend on code that uses `--guardedness`, this is why I do not want to declare it.  I happens to import a module (`Data.Container.Indexed.FreeMonad`) that imports another module (`Data.Container.Indexed`) that uses a function defined by guardedness  imported from a third module somewhere.

So now I have to declare that I use a broken feature (`--guardedness`) even though I don't, only because some module depended on more than it needed.

I would only want to declare `--guardedness` if I actually *used* a function exploiting guardedness, so the module level is imo the wrong place to track language flavours; it is too crude.  It should be on the definition level.

@jespercockx : It is simply that in the last (in our case) months before the release there shouldn't be backwards-incompatible changes since the impact takes longer to surface.  Such changes are best made immediately after the release.  The last months are best spent on eliminating regressions and uncontroversial fixes.

**Comment by jespercockx** [comment]
@andreasabel Yes, as I said I agree with the ideal that backwards incompatible changes should not be made too close to the release. But in this case @nad brought up a reasonable concern, namely that removing --safe from a module with --sized-types implicitly also turns on --guardedness, which is incompatible with --sized-types. In other words, there was going to be some kind of backwards incompatibility no matter what. So **doing nothing was not an option** (or at least not a good one). This is why I asked you to give feedback on the different possible solutions that I could think of:

1. disabling just --sized-types by default but disabling --guardedness when --sized-types is turned on (this required a lot of --no-guardedness flags in the stdlib)
2. disabling just --guardedness by default but disabling --sized-types when --guardedness is turned on (this required a lot of --no-sized-types flags in the stdlib)
3. keeping the status quo (and hence exposing people who used to have --safe to the unsound combination of --sized-types and --guardedness)
4. (proposed by @nad) disabling both --sized-types and --guardedness by default (this had the least overall impact)
 
You have still not indicated which solution would have had your preference.

**Comment by nad** [comment]
> I would only want to declare `--guardedness` if I actually _used_ a function exploiting guardedness, so the module level is imo the wrong place to track language flavours; it is too crude. It should be on the definition level.

Are you proposing to put flags like `--guardedness` before every definition? Or are you suggesting that Agda should somehow keep track of which features are "actually used" in a certain definition? Is that feasible?

**Comment by andreasabel** [comment]
> Or are you suggesting that Agda should somehow keep track of which features are "actually used" in a certain definition? Is that feasible?

Why not?  We also check which free variables are actually used in a definition, as opposed to just assuming that all variables that are in scope are used. :P

Checking termination is a relatively expensive operation, so setting a flag whether we used guardedness or not should be comparatively cheap.  Accumulating in a definition whether one of its references uses guardedness should also be comparatively cheap.  If we don't want to do this while type-checking: We already have functionality to compute all defined names used in some artifact, so we could `foldMap` over this list to determine whether any used guardedness.  The only effort is maybe needed for mutual recursion, where we need some saturation process, but the fixpoint should be reached after 2 iterations already.

I think if we keep track of used language variants per definition, we can contain the pandemic caused by infective options better...

**Comment by jespercockx** [comment]
I actually started drafting such a thing for the (proper) fix of #4908 - you can find my WIP on https://github.com/jespercockx/agda/tree/UnsafeThings. The main idea is that we store a data structure that records all "unsafe things" that are used in a file in the interface. That way we can easily check whether an unsafe feature is actually used or not without requiring an explicit flag. I agree with Andreas that this does not seem like it would cause performance issues, but I haven't tested it yet.

**Comment by nad** [comment]
> Why not? We also check which free variables are actually used in a definition, as opposed to just assuming that all variables that are in scope are used. :P

I see two potential problems:

* If one can't easily get information about whether a certain feature is used without running the typechecker twice, then this approach could be very expensive: I don't want to typecheck my code 2^n times. Does the termination checker produce a certificate from which one can tell if `--guardedness` is absolutely necessary to accept a certain definition (or could it easily produce such a certificate)? Is it easy to tell if `--cumulativity` is necessary?

* Code might become brittle. Consider a library that declares that it uses feature X, but does not use feature X in the definition `f`. If `f` is then changed so that it does use feature X, then code that uses `f` might break, even if this code does not depend on the computational behaviour of `f`.

**Comment by andreasabel** [comment]
https://github.com/agda/agda-stdlib/issues/1509

**Comment by andreasabel** [comment]
Zoom discussion 2021-06-03:

Fundamental issues:

- Flags like `--guardedness` are overloaded:
   * Turn on guardedness checker in current module.
   * Import a module that uses (transitively) guardedness.
   
- No warning about unused flags: inherited `--guardedness` may become obsolete as someone upstream ditches it, but Agda does not warn about it, so infection continues.

Current problematic cases:

- You have to declare `--guardedness` only because you import some library module that declares `--guardedness` (even though you make nor use definitions by guardedness).
- Libraries that aim for maximal generality have to be restructured to separate out parts that _actually_ use guardedness.

Alternatives discussed:

- Keeping `--guardedness` by default:  `--no-guardedness` would be coinfective, and the std-lib would have to declare `--no-guardedness` everywhere.
- Enabling either `--guardedness` or `--sized-types` by default and disabling it when the other is enabled: requires a log of `-no-...` flags.
- Not making `--guardedness` infective: Cannot trace it.
- Only infect `--safe` modules: Don't have the infrastructure for this.  With `--sized-types` and `--safe`, would have to declare `--no-guardedness` to stay safe, but this is not discoverable (one could maybe change the error message).  Complicates the flag logic even further.
-

**Comment by andreasabel** [comment]
I still do not agree with the current state of affairs.  
The way infective options are currently handled by Agda breaks important abstraction principles.
- You cannot use coinduction in a library without pushing either `--sized-types` or `--guardedness` too all your users and users' users etc. forever downstream.  In a large development that pulls in all kinds of libraries which again rest on libraries, the probability to having to declare an infective option will approximate 1.  
- You cannot do a strict "prelude" style organization of your project where you have one `Prelude` module that pulls in all your external dependencies and then all your project modules just import this `Prelude`.  If you happen to use coinduction somewhere, and thus import `Size` in `Prelude`, all your modules have to declare `--sized-types`.

I think the current design is biased badly towards users that want `--safe` and not acceptable for the reasons given above.

I also think I would not use Haskell if I had to declare e.g. `{-# LANGUAGE RankNTypes #-}` in all my projects only because somewhere upstream some library developer decided to use it in one of their modules.

**Comment by nad** [comment]
I added the `wontfix` label because we decided not to make any further changes before the release of Agda 2.6.2, not to suggest that the current design is frozen in stone forever.
